django/db/migrations/state.py

```diff
@@ -98,16 +98,31 @@ def __init__(self, models=None, real_apps=None):
         self.real_apps = real_apps
         self.is_delayed = False
         # {remote_model_key: {model_key: [(field_name, field)]}}
-        self.relations = None
+        self._relations = None
+
+    @property
+    def relations(self):
+        if self._relations is None:
+            self.resolve_fields_and_relations()
+        return self._relations
 
     def add_model(self, model_state):
-        app_label, model_name = model_state.app_label, model_state.name_lower
-        self.models[(app_label, model_name)] = model_state
+        model_key = model_state.app_label, model_state.name_lower
+        self.models[model_key] = model_state
+        if self._relations is not None:
+            self.resolve_model_relations(model_key)
         if 'apps' in self.__dict__:  # hasattr would cache the property
-            self.reload_model(app_label, model_name)
+            self.reload_model(*model_key)
 
     def remove_model(self, app_label, model_name):
         del self.models[app_label, model_name]
+        if self._relations is not None:
+            removed_model_key = (app_label, model_name)
+            self._relations.pop(removed_model_key, None)
+            for model_key, model_relations in list(self._relations.items()):
```

Comment:
I don't think you need `list()` here.

Reply:
`list()` is necessary because we're changing `self._relations` in this loop:
```
RuntimeError: dictionary changed size during iteration
```

Reply:
Ah, that is probably worth a comment, then. It's not obvious.

Reply:
Added.